GH-10090: Add `AmqpClientInboundGateway`

artembilan opened the pull request:
Related to: https://github.com/spring-projects/spring-integration/issues/10090

* Add `AmqpClientInboundGateway` that is mostly a copy/paste of the `AmqpClientMessageProducer`, but adds a reply-producing logic
* Cover with tests and document this new component
* Fix a couple typos in the `amqp-1.0.adoc`

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-10090: Add `AmqpClientInboundGateway`

Related to: https://github.com/spring-projects/spring-integration/issues/10090

* Add `AmqpClientInboundGateway` that is mostly a copy/paste of the `AmqpClientMessageProducer`,
but adds a reply-producing logic
* Cover with tests and document this new component
* Fix a couple typos in the `amqp-1.0.adoc`

artembilan commented:
After doing this, I don't see a value in the `AmqpClientMessageSource`: who would need to poll messages from the queue on demand? 🤷 

The `RabbitAmqpTemplate.receive()` API could be turned into a `MessageSource` very easy:
```
@Bean 
IntegrationFlow amqpClientPollingFlow(RabbitAmqpTemplate rabbitAmqpTemplate) {
    return IntegrationFlow.fromSupplier(rabbitAmqpTemplate::receive)
                          .handle((p, h) -> p, e -> e.async(true))
                          ...
}
```

So, the point is that `RabbitAmqpTemplate.receive()` returns a `CompletableFuture<Message>` which we don't want to `join()` immediately.

cppwfs commented on src/reference/antora/modules/ROOT/pages/amqp/amqp-1.0.adoc:
```diff
@@ -89,3 +89,29 @@ AmqpClientMessageProducer batchAmqpClientMessageProducer(AmqpConnectionFactory c
+For automatic replies correlation with their requests, a `replyTo` property of the request message must be supplied.
+For example, the `RabbitAmqpTemplate.sendAndReceive()` relies on the `RpcClient` from RabbitMQ AMQP 1.0 library which generates an exclusive and auto-deleted queue.
+Alternatively, the reply address could be set as a `replyExchange` and `replyRoutingKey` (optional) or `replyQueue` on the `AmqpClientInboundGateway` which are delegated to the `RabbitAmqpTemplate` default options.
+The `messageId` or `correllationId` request message properties can be used for associating with replies.
```
correllationId --> correlation

cppwfs commented on src/reference/antora/modules/ROOT/pages/amqp/amqp-1.0.adoc:
```diff
@@ -89,3 +89,29 @@ AmqpClientMessageProducer batchAmqpClientMessageProducer(AmqpConnectionFactory c
+
+For automatic replies correlation with their requests, a `replyTo` property of the request message must be supplied.
+For example, the `RabbitAmqpTemplate.sendAndReceive()` relies on the `RpcClient` from RabbitMQ AMQP 1.0 library which generates an exclusive and auto-deleted queue.
+Alternatively, the reply address could be set as a `replyExchange` and `replyRoutingKey` (optional) or `replyQueue` on the `AmqpClientInboundGateway` which are delegated to the `RabbitAmqpTemplate` default options.
```
Do we want to mention it can't be both?

cppwfs reviewed (changes requested):
Looks great!
Just a couple of doc questions

artembilan committed:
Fix docs typos

cppwfs reviewed (approved):
Sweet!
 
 I'll merge once it successfully builds.

artembilan commented:
Hold on from merging, please.

cppwfs commented:
On hold till I get the go ahead.

artembilan committed:
Extract the `IntegrationRabbitAmqpMessageListener` as a top-level (package private) class.
Serves as a reusable unit of work for both `AmqpClientMessageProducer` & `AmqpClientInboundGateway`.
The one-way and request-reply parts are handled as a `BiConsumer` action injection
into this `IntegrationRabbitAmqpMessageListener` instance.
Now both `AmqpClientMessageProducer` & `AmqpClientInboundGateway` are much simpler.

cppwfs reviewed (changes requested):
Great job on consolidating the code.  The `red` is always pleasing to see!   :-)

Just a couple of nitpicks.

artembilan committed:
* Improve `AmqpClientMessageProducer.processRequest` Javadoc
mentioning that `requestMessage` param is out of use
* Be more specific with the type for `UnsupportedOperationException`
in the `IntegrationRabbitAmqpMessageListener`

cppwfs reviewed (approved):
LGTM!

This is Awesome!
